content/posts/2023-01-16-react-native-testing/index.md

```diff
@@ -0,0 +1,392 @@
+
+```js
+test('should render the label of the social link', () => {
+  const { getByText } = render(
+    <SocialLinks
+      type="twitter"
+      label="John Doe's Twitter"
+      link="https://twitter.com/"
+    />,
+  );
+
+  const label = getByText("John Doe's Twitter 123");
+  expect(label).toBeTruthy();
+});
+```
+
+![react-native-testing-test-1-fail.png](./images/test-1-fail.png)
+
+Why did the test fail? Because in our test function we are expecting the label to be "John Doe's Twitter 123" but in the component we are passing the label as "John Doe's Twitter".
+
+## Testing the second case : The user will see the icon of the social link
+
+Let's now test the second case that the user will see the icon of the social link. Let's add the following code to the test file:
+
+```js
+test('should render the icon of the social link', () => {
+  const { getByRole } = render(
+    <SocialLinks
+      type="twitter"
+      label="John Doe's Twitter"
+      link="https://twitter.com/"
+    />,
+  );
+
+  const icon = getByRole('image');
+  expect(icon).toBeTruthy();
+});
+```
+
+Again, let's break down the code. We are using the `getByRole` method to get the icon of the social link. Why `getByRole`? Because we have set the `accessibilityRole` to `image` in the component. We are then using the same `toBeTruthy` matcher to check if the icon is present or not. If the icon is present, the test will pass. If the icon is not present, the test will fail.
+
+If you run the test now, you will see the following output:
+
+![react-native-testing-test-2.png](./images/test-2-pass.png)
+
+Great, we have now made sure that our SocialLinks component is rendering the label and the icon of the social link.
+
+## Testing the third and fourth case : Clicking on the social link should open the correct link
+
+Now onto our third & fourth case. We will be testing if users go to twitter.com when they click on the twitter link. Let's add the following code to the test file:
+
+```js
+import { Linking } from 'react-native'; // remember to import Linking
+
+test('should open the twiiter link when the twitter link is pressed', () => {
+  const { getByText } = render(
+    <SocialLinks
+      type="twitter"
+      label="John Doe's Twitter"
+      link="https://twitter.com/"
+    />,
+  );
+
+  const label = getByText("John Doe's Twitter");
+  fireEvent.press(label);
+  expect(Linking.openURL).toHaveBeenCalledWith('https://twitter.com/');
+});
+```
+
+Similarly to the previous test cases, let's break down the code. We are using the `fireEvent` method to fire the `press` event on the label. We are then using the `toHaveBeenCalledWith` matcher to check if the `Linking.openURL` method is called with the correct link. If the link is correct, the test will pass. If the link is not correct, the test will fail.
```

Comment:
Maybe it's a good idea to remind people that the SocialLinks component, uses `Linking.openURL` when users click on it, that's why in our test, we are making sure that it was actually called. 

I think the confusion will be, "What does Linking has to do with this test", or "Do we always use Linking to test button clicks?"

Reply:
Added why.